File: modules/rust/scaffold/packages/domain/src/repositories/post_repository.rs
use async_trait::async_trait;

use crate::errors::RepositoryResult;
use crate::models::post::Post;

#[async_trait]
pub trait PostRepository: Send + Sync {
    async fn find_all(&self) -> RepositoryResult<Vec<Post>>;
    async fn find_by_id(&self, id: &str) -> RepositoryResult<Post>;
    async fn create(&self, post: &Post) -> RepositoryResult<Post>;
    async fn update(&self, post: &Post) -> RepositoryResult<Post>;
    async fn delete(&self, id: &str) -> RepositoryResult<()>;
}

File: templates/effect-api/scaffold/apps/{{app_name}}/src/domain/repositories/post-repository.ts
import { Context, Data, type Effect, type Option, Schema } from "effect"
import type { Post } from "@/domain/models/post"

// Schema.TaggedError (not Data.TaggedError) so it doubles as the HttpApi error schema (404).
export class PostNotFound extends Schema.TaggedError<PostNotFound>()("PostNotFound", {
	id: Schema.String,
}) {}

export class DbError extends Data.TaggedError("DbError")<{
	readonly cause: unknown
}> {}

export type NewPost = {
	readonly id: string
	readonly title: string
	readonly body: string
	readonly publishedAt: Option.Option<Date>
}

export type PostPatch = {
	readonly title: Option.Option<string>
	readonly body: Option.Option<string>
}

export class PostRepository extends Context.Tag("PostRepository")<
	PostRepository,
	{
		readonly findAll: () => Effect.Effect<ReadonlyArray<Post>, DbError>
		readonly findById: (id: string) => Effect.Effect<Option.Option<Post>, DbError>
		readonly create: (input: NewPost) => Effect.Effect<Post, DbError>
		readonly update: (id: string, patch: PostPatch) => Effect.Effect<Post, DbError | PostNotFound>
		readonly delete: (id: string) => Effect.Effect<void, DbError | PostNotFound>
	}
>() {}
